File: scripts/smoke.mjs
/**
 * Runs the app against every fixture and says whether it survived.
 *
 *   node scripts/smoke.mjs [--release] [--keep]
 *
 * The app does the checking; this starts it, holds the clock, and reads the
 * verdict. Three things live out here and not in the app:
 *
 * **The deadline.** Two of the defect classes this exists to catch are the
 * event loop failing to turn, and a timer inside that loop would stop with it.
 * A process that stops answering is killed from outside, and a results file
 * with no summary line is how that is told apart from a run that failed.
 *
 * **The second process.** The single-instance hand-off needs two of them, and
 * an app cannot start itself.
 *
 * **Finding the binary.** Which build is under test is the runner's business —
 * and the release one is the point, because that is where the crashes this
 * targets only ever appeared.
 */
import { spawn } from "node:child_process";
import { mkdtemp, readFile, rm } from "node:fs/promises";
import { existsSync } from "node:fs";
import { tmpdir } from "node:os";
import path from "node:path";

const release = process.argv.includes("--release");
const keep = process.argv.includes("--keep");
const root = process.cwd();
const manifest = path.join(root, "fixtures/smoke.json");

/**
 * Where the binary is, in the order worth looking.
 *
 * More than one place because the macOS release is built for
 * `universal-apple-darwin`, which cargo puts under the target's own directory
 * rather than the profile's. `DVIEWER_EXE` wins over both, for anyone testing
 * something that was built elsewhere — an installed copy, say.
 */
const name = process.platform === "win32" ? "dviewer.exe" : "dviewer";
const profile = release ? "release" : "debug";
const candidates = [
  process.env.DVIEWER_EXE,
  path.join(root, "src-tauri/target", profile, name),
  path.join(root, "src-tauri/target/universal-apple-darwin", profile, name),
].filter(Boolean);
const exe = candidates.find((candidate) => existsSync(candidate));

/** The whole sweep. Generous: a cold runner opening a 500MB fixture is slow. */
const SWEEP_TIMEOUT_MS = 10 * 60_000;
/** One hand-off. If it has not arrived by now it is not going to. */
const HANDOFF_TIMEOUT_MS = 60_000;
/**
 * How long to wait for the listening process to say it is listening.
 *
 * It says so rather than being assumed ready after a pause. The request is
 * handed over as an event, and an event nobody is listening for yet is lost
 * without a sound — so a pause here is a guess about webview boot time, and the
 * guess is wrong on exactly the machine that matters: a cold CI runner starting
 * a release build under a virtual display.
 */
const READY_TIMEOUT_MS = 60_000;

const sleep = (ms) => new Promise((resolve) => setTimeout(resolve, ms));

function fail(message) {
  console.error(`  ✗ ${message}`);
  process.exitCode = 1;
}

/** Start the app and wait for it to end, or kill it when the clock runs out. */
function run(args, timeoutMs) {
  return new Promise((resolve) => {
    const child = spawn(exe, args, { stdio: ["ignore", "pipe", "pipe"], windowsHide: false });
    let stderr = "";
    child.stderr?.on("data", (chunk) => (stderr += chunk));

    const timer = setTimeout(() => {
      child.kill("SIGKILL");
      resolve({ code: null, killed: true, stderr });
    }, timeoutMs);

    child.on("error", (err) => {
      clearTimeout(timer);
      resolve({ code: null, killed: false, stderr: String(err) });
    });
    child.on("close", (code) => {
      clearTimeout(timer);
      resolve({ code, killed: false, stderr });
    });
  });
}

/**
 * Read a results file.
 *
 * The summary line is the completion mark. Its absence means the process never
 * got to the end, whatever else the file says — and the last line before it is
 * then the document it was working on.
 */
async function results(file) {
  const text = await readFile(file, "utf8").catch(() => "");
  const lines = text
    .split("\n")
    .filter((line) => line.trim() !== "")
    .map((line) => JSON.parse(line));
  const summary = lines.at(-1)?.summary ?? null;
  return { lines: summary ? lines.slice(0, -1) : lines, summary };
}

/**
 * Wait until the listening process has said it is listening.
 *
 * `gone` means it ended before it got there, which on this check means
 * something else already held the single-instance lock.
 */
async function waitForListening(file, ended) {
  const deadline = Date.now() + READY_TIMEOUT_MS;
  while (Date.now() < deadline) {
    const { lines } = await results(file);
    if (lines.some((line) => line.step === "listening")) return "ready";
    if (ended()) return "gone";
    await sleep(100);
  }
  return "timeout";
}

function report(lines) {
  for (const line of lines) {
    // Only `ok` decides. A line can carry an error and still have passed — an
    // archive whose one document was refused shows its list and says why, and
    // that banner is the correct outcome rather than a failure.
    if (line.ok !== false) continue;
    fail(
      `${line.file ?? line.step}: ${line.stage ?? ""} ${line.error ?? ""}`.trim() +
        (line.view ? ` (${line.view} 로 열림, ${line.expect} 를 기대)` : ""),
    );
  }
}

// --- go ----------------------------------------------------------------------

if (!exe) {
  console.error("빌드된 앱을 찾지 못했습니다. 찾아본 곳:");
  for (const candidate of candidates) console.error(`  ${candidate}`);
  console.error(
    release
      ? "  npm run build && cd src-tauri && cargo build --release --features custom-protocol"
      : "  npm run build && cd src-tauri && cargo build --features custom-protocol",
  );
  process.exit(2);
}
if (!existsSync(manifest)) {
  console.error(`픽스처가 없습니다. 먼저: node scripts/gen-fixtures.mjs`);
  process.exit(2);
}

const work = await mkdtemp(path.join(tmpdir(), "dviewer-smoke-"));
console.log(`스모크 (${release ? "릴리스" : "디버그"} 빌드)`);

// 1 — every fixture, through the ordinary open pipeline.
{
  const out = path.join(work, "sweep.jsonl");
  const started = Date.now();
  const ended = await run([`--smoke=${manifest}`, `--smoke-out=${out}`], SWEEP_TIMEOUT_MS);
  const { lines, summary } = await results(out);

  if (ended.code === 2) fail(`하네스가 시작하지 못했습니다: ${ended.stderr.trim()}`);
  else if (!summary) {
    const last = lines.at(-1);
    fail(
      ended.killed
        ? `${Math.round(SWEEP_TIMEOUT_MS / 1000)}초 안에 끝나지 않았습니다` +
            (last ? ` — 마지막으로 연 것: ${last.file}` : " — 아무것도 열지 못했습니다")
        : `끝까지 가지 못했습니다 (종료 코드 ${ended.code})` +
            (last ? ` — 마지막으로 연 것: ${last.file}` : ""),
    );
  } else {
    report(lines);
    if (summary.failed > 0) fail(`${summary.total}개 중 ${summary.failed}개 실패`);
    else console.log(`  ✓ 픽스처 ${summary.total}개, ${Math.round((Date.now() - started) / 1000)}초`);
  }
}

// 2 — the single-instance hand-off. A second `dviewer` must give its arguments
//     to the first and exit; only the first can say they arrived.
for (const [label, extra] of [
  ["단일 인스턴스 전달", []],
  ["--new 창", ["--new"]],
]) {
  const out = path.join(work, `${extra.length ? "new" : "handoff"}.jsonl`);
  let listenerEnded = false;
  const listening = run(["--smoke-listen", `--smoke-out=${out}`], HANDOFF_TIMEOUT_MS).then(
    (ended) => {
      listenerEnded = true;
      return ended;
    },
  );

  const ready = await waitForListening(out, () => listenerEnded);
  if (ready === "gone") {
    // The listener holds the single-instance lock, so it must still be running.
    // If it is not, something else already held it — an open dviewer — and this
    // check would be measuring that instead.
    fail(`${label}: dviewer 가 이미 떠 있습니다. 닫고 다시 돌려 주세요.`);
    continue;
  }
  if (ready === "timeout") {
    fail(`${label}: 듣는 프로세스가 ${READY_TIMEOUT_MS / 1000}초 안에 준비되지 않았습니다`);
    continue;
  }

  const second = await run([...extra, path.join(root, "fixtures/sample.md")], HANDOFF_TIMEOUT_MS);
  if (second.killed) fail(`${label}: 두 번째 프로세스가 스스로 종료하지 않았습니다`);

  const first = await listening;
  const { lines, summary } = await results(out);
  report(lines);

  if (!summary) fail(`${label}: 첫 프로세스가 전달을 보고하지 않았습니다`);
  else if (first.code !== 0 || summary.failed > 0) fail(`${label}: 전달이 확인되지 않았습니다`);
  else console.log(`  ✓ ${label}`);
}

if (!keep) await rm(work, { recursive: true, force: true });
else console.log(`  결과: ${work}`);

if (process.exitCode) console.error("스모크 실패");
else console.log("스모크 통과");

File: src-tauri/src/smoke.rs
//! The self-check: open everything, and say what happened.
//!
//! This exists because of what the unit layers cannot reach. Every defect this
//! repository has shipped past its tests was of one class — the app started,
//! something was opened, and it fell over: a crash only the release profile
//! produced, an event loop that blocked on itself, a capability that was never
//! wired. None of them needed a complicated interaction to reproduce. What they
//! needed was a real window, a real IPC round trip and a real webview boot.
//!
//! So the harness is the app. It runs the ordinary open pipeline over a list of
//! documents and records how far each got. There is no driver, no selector and
//! no timing to flake on — and it works on the *released* binary, which is the
//! only way to catch something that only happens there.
//!
//! Two rules shape everything here.
//!
//! **Results are written a line at a time and flushed.** A single document
//! assembled at the end would be empty exactly when this harness earns its
//! keep: if the process dies, the file is all that is left, and the last line
//! written names what killed it.
//!
//! **Nothing here imposes a deadline.** The runner outside does. Two of the
//! three defect classes above are the event loop failing to turn, and a timer
//! living inside that loop would stop with it.

use std::fs::File;
use std::io::Write;
use std::path::{Path, PathBuf};
use std::time::Instant;

use parking_lot::Mutex;
use serde::{Deserialize, Serialize};

use crate::cli::{Smoke, SmokeMode};

/// Everything went as the plan said it would.
pub const OK: i32 = 0;
/// At least one document did not. Which ones is in the results file.
pub const MISMATCH: i32 = 1;
/// The harness itself could not run — no manifest, or nowhere to write.
pub const BROKEN: i32 = 2;

/// One document to open, and what should become of it.
#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Step {
    /// As the manifest names it, which is what a failure is reported under.
    pub file: String,
    /// Absolute, resolved against the manifest's own directory — the manifest
    /// sits with the fixtures, so it is the only thing that knows where they
    /// are.
    #[serde(default)]
    pub path: String,
    /// The view this should end in, or `error` when refusing is the right
    /// answer. One field, and it is what separates an archive that unwraps to
    /// its single document from one that falls back to its list.
    pub expect: String,
    /// A second thing to do once it is open, for the paths a file alone does
    /// not reach: `openEntry`, `toggleHeader`.
    #[serde(default)]
    pub then: Option<String>,
}

/// A self-check in progress.
pub struct SmokeRun {
    plan: Vec<Step>,
    out: Mutex<File>,
    started: Instant,
    tally: Mutex<(usize, usize)>,
    /// The window whose destruction ends this run, if one has asked.
    ///
    /// The `--new` half of the round trip finishes by closing its own window,
    /// because the thing being checked is what happens *after* that: a window
    /// going away has to take its documents with it, and no unit test can
    /// reach that code — it lives in a Tauri event handler. So the run cannot
    /// end when the frontend is done; it ends when the window is really gone.
    finish_when_gone: Mutex<Option<String>>,
}

impl SmokeRun {
    /// Read the plan and open the results file, or say why not.
    ///
    /// Failing here is `BROKEN` rather than `MISMATCH`: nothing about the app
    /// has been tested yet, and reporting a document as broken when the harness
    /// could not read its own manifest would be a lie in the direction that
    /// wastes the most time.
    pub fn start(smoke: &Smoke) -> Result<Self, String> {
        let plan = match &smoke.mode {
            SmokeMode::Listen => Vec::new(),
            SmokeMode::Run { manifest } => read_plan(manifest)?,
        };
        let out = File::create(&smoke.out)
            .map_err(|e| format!("cannot write {}: {e}", smoke.out.display()))?;
        Ok(Self {
            plan,
            out: Mutex::new(out),
            started: Instant::now(),
            tally: Mutex::new((0, 0)),
            finish_when_gone: Mutex::new(None),
        })
    }

    pub fn plan(&self) -> &[Step] {
        &self.plan
    }

    /// Append one result and flush it.
    ///
    /// Flushed rather than buffered, and the reason is the whole design: this
    /// file is the only thing that survives a process that stops existing.
    ///
    /// `ok` is folded into the line rather than kept beside it. Whoever reads
    /// this file later has only the file, and a result that does not say
    /// whether it passed makes them reconstruct the verdict from the fields —
    /// which is exactly the kind of guessing that reads a banner as a failure.
    /// Done here rather than in the command, because not every line comes from
    /// the frontend: the destroy handler writes one too.
    pub fn record(&self, line: serde_json::Value, ok: bool) {
        {
            let mut tally = self.tally.lock();
            tally.0 += 1;
            if !ok {
                tally.1 += 1;
            }
        }
        let mut line = line;
        if let Some(object) = line.as_object_mut() {
            object.insert("ok".into(), ok.into());
        }
        let mut out = self.out.lock();
        let _ = writeln!(out, "{line}");
        let _ = out.flush();
    }

    /// End the run when `window` is destroyed rather than now.
    pub fn finish_when_gone(&self, window: &str) {
        *self.finish_when_gone.lock() = Some(window.to_owned());
    }

    /// Whether the destruction of `window` is what ends the run.
    pub fn ends_with(&self, window: &str) -> bool {
        self.finish_when_gone.lock().as_deref() == Some(window)
    }

    /// Write the summary and say how the process should end.
    ///
    /// The summary line is also the completion mark: a results file without one
    /// is a run that did not finish, whatever its last line says. The runner
    /// outside reads it that way, which is what lets a killed process be told
    /// apart from a failing one.
    pub fn finish(&self) -> i32 {
        let (total, failed) = *self.tally.lock();
        let summary = serde_json::json!({
            "summary": {
                "total": total,
                "failed": failed,
                "ms": self.started.elapsed().as_millis() as u64,
            }
        });
        let mut out = self.out.lock();
        let _ = writeln!(out, "{summary}");
        let _ = out.flush();
        if failed == 0 { OK } else { MISMATCH }
    }
}

fn read_plan(manifest: &Path) -> Result<Vec<Step>, String> {
    let text = std::fs::read_to_string(manifest)
        .map_err(|e| format!("cannot read {}: {e}", manifest.display()))?;
    let mut plan: Vec<Step> = serde_json::from_str(&text)
        .map_err(|e| format!("cannot parse {}: {e}", manifest.display()))?;
    if plan.is_empty() {
        return Err(format!("{} lists nothing to open", manifest.display()));
    }

    let beside = manifest.parent().unwrap_or_else(|| Path::new("."));
    for step in &mut plan {
        step.path = PathBuf::from(beside)
            .join(&step.file)
            .to_string_lossy()
            .into_owned();
    }
    Ok(plan)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn wrote(name: &str, body: &str) -> PathBuf {
        let path = std::env::temp_dir().join(format!("dviewer-smoke-{name}"));
        std::fs::create_dir_all(&path).expect("dir");
        let manifest = path.join("smoke.json");
        std::fs::write(&manifest, body).expect("write");
        manifest
    }

    /// The manifest sits with the fixtures, so it is what says where they are.
    /// A harness run from anywhere else must still find them.
    #[test]
    fn a_document_is_found_beside_the_manifest_that_lists_it() {
        let manifest = wrote("resolve", r#"[{"file":"sample.md","expect":"prose"}]"#);
        let plan = read_plan(&manifest).expect("plan");
        assert_eq!(plan[0].file, "sample.md");
        assert!(plan[0].path.ends_with("sample.md"));
        assert!(
            PathBuf::from(&plan[0].path).is_absolute() || plan[0].path.contains(std::path::MAIN_SEPARATOR),
            "resolved beside the manifest, not left as a bare name: {}",
            plan[0].path
        );
    }

    #[test]
    fn a_follow_up_step_is_optional() {
        let manifest = wrote(
            "then",
            r#"[{"file":"a.zip","expect":"archive","then":"openEntry"},
                {"file":"b.md","expect":"prose"}]"#,
        );
        let plan = read_plan(&manifest).expect("plan");
        assert_eq!(plan[0].then.as_deref(), Some("openEntry"));
        assert_eq!(plan[1].then, None);
    }

    /// Every one of these is the harness being unable to run, not a document
    /// being wrong — and the two must not be reported as the same thing.
    #[test]
    fn a_manifest_that_cannot_be_used_says_so_rather_than_failing_a_document() {
        assert!(read_plan(Path::new("nowhere/smoke.json")).is_err());
        assert!(read_plan(&wrote("garbage", "not json at all")).is_err());
        assert!(read_plan(&wrote("empty", "[]")).is_err());
    }

    /// The summary is the completion mark. Without it the runner outside cannot
    /// tell a process that failed from one that was killed.
    #[test]
    fn the_results_end_with_a_summary_that_counts_what_went_wrong() {
        let manifest = wrote("finish", r#"[{"file":"a.md","expect":"prose"}]"#);
        let out = manifest.with_file_name("out.jsonl");
        let run = SmokeRun::start(&Smoke {
            mode: SmokeMode::Run {
                manifest: manifest.clone(),
            },
            out: out.clone(),
        })
        .expect("start");

        run.record(serde_json::json!({"file": "a.md", "ok": true}), true);
        run.record(serde_json::json!({"file": "b.md", "ok": false}), false);
        assert_eq!(run.finish(), MISMATCH);

        let written = std::fs::read_to_string(&out).expect("read");
        let lines: Vec<&str> = written.lines().collect();
        assert_eq!(lines.len(), 3, "two results and a summary");
        let summary: serde_json::Value = serde_json::from_str(lines[2]).expect("summary");
        assert_eq!(summary["summary"]["total"], 2);
        assert_eq!(summary["summary"]["failed"], 1);
    }

    #[test]
    fn a_run_with_nothing_wrong_ends_at_zero() {
        let manifest = wrote("clean", r#"[{"file":"a.md","expect":"prose"}]"#);
        let run = SmokeRun::start(&Smoke {
            mode: SmokeMode::Run { manifest: manifest.clone() },
            out: manifest.with_file_name("clean.jsonl"),
        })
        .expect("start");
        run.record(serde_json::json!({"file": "a.md", "ok": true}), true);
        assert_eq!(run.finish(), OK);
    }


    /// Every line says whether it passed, whoever wrote it.
    ///
    /// The frontend used to fold `ok` in on its way through the command, which
    /// left the one line the backend writes by itself — the reclaim line from
    /// the destroy handler — without a verdict. Whoever reads this file has
    /// only the file.
    #[test]
    fn a_recorded_line_carries_its_own_verdict() {
        let manifest = wrote("verdict", r#"[{"file":"a.md","expect":"prose"}]"#);
        let out = manifest.with_file_name("verdict.jsonl");
        let run = SmokeRun::start(&Smoke {
            mode: SmokeMode::Run { manifest: manifest.clone() },
            out: out.clone(),
        })
        .expect("start");

        run.record(serde_json::json!({"step": "reclaim", "window": "doc-1", "docs": 1}), true);
        run.record(serde_json::json!({"step": "reclaim", "window": "doc-2", "docs": 0}), false);
        run.finish();

        let written = std::fs::read_to_string(&out).expect("read");
        let lines: Vec<serde_json::Value> = written
            .lines()
            .map(|line| serde_json::from_str(line).expect("json"))
            .collect();
        assert_eq!(lines[0]["ok"], true);
        assert_eq!(lines[0]["docs"], 1);
        assert_eq!(lines[1]["ok"], false, "nothing reclaimed is a failure");
    }

    /// Which window's death ends the run, and only that one's.
    #[test]
    fn a_run_can_be_told_to_end_when_a_window_goes() {
        let manifest = wrote("gone", r#"[{"file":"a.md","expect":"prose"}]"#);
        let run = SmokeRun::start(&Smoke {
            mode: SmokeMode::Run { manifest: manifest.clone() },
            out: manifest.with_file_name("gone.jsonl"),
        })
        .expect("start");

        assert!(!run.ends_with("doc-1"), "nothing ends the run until something asks");
        run.finish_when_gone("doc-1");
        assert!(run.ends_with("doc-1"));
        assert!(!run.ends_with("main"), "another window closing is not the end");
    }

    /// Listening opens nothing, so it has no plan — and must still start.
    #[test]
    fn listening_starts_without_a_manifest() {
        let out = std::env::temp_dir().join("dviewer-smoke-listen.jsonl");
        let run = SmokeRun::start(&Smoke {
            mode: SmokeMode::Listen,
            out,
        })
        .expect("start");
        assert!(run.plan().is_empty());
    }
}
